tests/entrypoints/openai/test_basic.py

```diff
@@ -59,3 +59,22 @@ async def test_log_metrics(client: openai.AsyncOpenAI):
     response = requests.get(base_url + "/metrics")
 
     assert response.status_code == HTTPStatus.OK
+
+
+@pytest.mark.asyncio
+async def test_get_readiness_ok(client: openai.AsyncOpenAI):
+    """Test the technical route /readiness when the model is fully loaded"""
+    base_url = str(client.base_url)[:-3].strip("/")
+
+    response = requests.get(base_url + "/ready")
+
+    assert response.status_code == HTTPStatus.OK
+
+@pytest.mark.asyncio
+async def test_get_readiness_ok(client: openai.AsyncOpenAI):
```

Comment:
I guess you're going to update this test to check when the server is *not* ready?

Reply:
yes exactly, I am working on it

Reply:
Regarding readiness, I have worked on create a proper unit test when the server is not ready. I was thinking to test that when the model weight were not loaded or KV cache was not set up, readiness endpoints would return an error message. 

But when I have checked how VLLM server was launched, I realized that the endpoints would not be callable until the server was properly deployed and model was loaded with KV cache setup. So I don't see how I can test if the model weights are loaded or KV cache is set up, because the endpoint for readiness will not be callable until the weights are loaded and KV cache is set up as well. 

So, do you have any other idea how to do this test? 

Is it compulsory to add this test for the PR to be merged?

Reply:
Hmm, I think the existing behavior would basically defeat the purpose of this PR in terms of fulfilling #6073. If the server cannot accept any requests until everything has been fully loaded, then there is essentially no difference between `/ready` and `/health`. Instead, we should enable the `/health` endpoint to respond before the vLLM engine has finished starting up.

Reply:
- > Hmm, this would basically defeat the purpose of this PR in terms of fulfilling #6073. If the server cannot accept any requests until everything has been fully loaded, then there is essentially no difference between `/ready` and `/health`. Instead, we should enable the `/health` endpoint to respond before the vLLM engine has finished starting up.

I understand your point. Furthermore, after having checked the new release 0.5.4, I have noticed several features has been added on the rpc server to check if it is ready. So, I don't think the readiness endpoint implemented in this PR is still useful. 

These are the next possible actions I propose : 

- Close this PR for two reasons : 
1. The readiness endpoint implemented in this PR checks if the model weights are loaded and kv cache set up. An audit of the code shows that the server can not be up until these weights et kv cache are properly loaded. So, there is not point to check these as the health endpoint will do it. 
2. As cited in https://github.com/vllm-project/vllm/issues/6073, regarding deployment on k8s, there is a need to implement k8s probes (startup, readiness, liveness) to have an autonomous deployment which will wait for model to be loaded and then marks the pod as ready when the health checkpoint return 200. I think this not directly related to the vllm server, but it is more about implementing proper helm chart which will configure it.

- I can open a new PR which proposes examples of helm chart for VLLM deployment on k8s and  allow to deploy a VLLM server on k8s, including k8s probes.

Reply:
I do have a deployment using a startup probe and liveness checks afterwards.

The main issue I have with the startup probe is that it is a workaround for applications with a potentially long startup unable to communicate their readiness. One never knows how much time a pod needs (downloading model, weights, ...) and in the meantime the liveness cannot be checked.


Switching to a liveness check that is available very early during startup would be nice, but that then requires a readiness indicator to not send traffic until vLLM is ready.

I have not looked at the recent changes yet. But there really should be a way to bring up the webserver (endpoint) early and also to indicate when it's ready.

Additionally I would love for some metrics to also be returned during the initialization phase, allowing for that to be observed.

Reply:
@frittentheke I have opened a PR #9199 to share a chart helm in order to have an example how to deploy vllm on k8s.